qiskit/transpiler/passes/optimization/commutation_checker.py

```diff
@@ -0,0 +1,137 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Code from commutative_analysis pass that checks commutation relations between DAG nodes."""
+
+from functools import lru_cache
+
+import numpy as np
+from qiskit.quantum_info.operators import Operator
+from qiskit.dagcircuit import DAGOpNode, DAGDepNode
+
+
+@lru_cache(maxsize=None)
+def _identity_op(num_qubits):
+    """Cached identity matrix"""
+    return Operator(
+        np.eye(2**num_qubits), input_dims=(2,) * num_qubits, output_dims=(2,) * num_qubits
+    )
+
+
+class CommutationChecker:
```

Comment:
These are all good ideas: removing the use of the dag node classes (handled in [a1284b9](https://github.com/Qiskit/qiskit-terra/pull/8184/commits/a1284b9fb9c33c1849fc1178c940f49219b7d009)) and moving the class to a more central place (still note sure where exactly to move it).

Reply:
As per the previous suggestion, the function `commute` in `commutation_checker` now looks as 
```
def commute(self, op1: Operation, qargs1: List, cargs1: List, op2: Operation, qargs2: List, cargs2: List)
```
In particular, it does not depend on `DAGNode` or `DAGDepNode` and can also be called on gates/operations in `QuantumCircuit`. 

After some discussion with @ShellyGarion and @eliarbel, I have moved `commutation_checker` to `qiskit/circuit`. This refactoring also allowed to avoid runtime imports.

Reply:
I guess the only question for me is why `qiskit/circuit` instead of `qiskit/quantum_info`? I could see a case for either location and don't feel to strongly one way or the other. I'm fine with this decision but having a comment in the PR review and/or as part of commit message for future reference might be useful in case someone attempts to revisit it's location again.

Reply:
I don't feel strongly one way or another either, but it seems that `quantum_info` should be agnostic to the encompassing circuit (i.e. it can reason about how to represent a Clifford using stabilizer/destabilizer tables or even about how to create a circuit for a Clifford, but it should not reason about where Clifford sits in the larger circuit), while commutation checker does look where the pair of gates sit in the larger circuit (looking at `qargs` and `cargs`). So personally `circuit` seems a better fit.